dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/resourcetable/table/OrganisationUnitGroupSetResourceTable.java

```diff
@@ -28,64 +28,100 @@
 
   public OrganisationUnitGroupSetResourceTable(
-      List<OrganisationUnitGroupSet> objects, int organisationUnitLevels, String tableType) {
-    super(objects);
+      List<OrganisationUnitGroupSet> groupSets, int organisationUnitLevels, Logged logged) {
+    this.groupSets = groupSets;
     this.organisationUnitLevels = organisationUnitLevels;
-    this.tableType = tableType;
+    this.logged = logged;
   }
 
   @Override
-  public ResourceTableType getTableType() {
-    return ResourceTableType.ORG_UNIT_GROUP_SET_STRUCTURE;
+  public Table getTable() {
+    return new Table(toStaging(TABLE_NAME), getColumns(), getPrimaryKey(), getIndexes(), logged);
   }
 
-  @Override
-  public String getCreateTempTableStatement() {
-    String statement =
-        "create "
-            + tableType
-            + " table "
-            + getTempTableName()
-            + " ("
-            + "organisationunitid bigint not null, "
-            + "organisationunitname varchar(230), "
-            + "startdate date, ";
-
-    for (OrganisationUnitGroupSet groupSet : objects) {
-      statement += quote(groupSet.getShortName()) + " varchar(230), ";
-      statement += quote(groupSet.getUid()) + " character(11), ";
+  private List<Column> getColumns() {
+    List<Column> columns =
+        Lists.newArrayList(
+            new Column("organisationunitid", DataType.BIGINT, Nullable.NOT_NULL),
+            new Column("organisationunitname", DataType.VARCHAR_255, Nullable.NOT_NULL),
+            new Column("startdate", DataType.DATE));
+
+    for (OrganisationUnitGroupSet groupSet : groupSets) {
+      columns.addAll(
+          List.of(
+              new Column(groupSet.getShortName(), DataType.VARCHAR_255),
+              new Column(groupSet.getUid(), DataType.CHARACTER_11)));
     }
 
-    return removeLastComma(statement) + ")";
+    return columns;
+  }
+
+  private List<String> getPrimaryKey() {
+    return List.of("organisationunitid");
+  }
+
+  public List<Index> getIndexes() {
+    return List.of(
+        new Index(
+            appendRandom("in_orgunitgroupsetstructure_not_null"),
+            IndexType.BTREE,
+            Unique.NON_UNIQUE,
+            List.of("organisationunitid", "startdate"),
+            "startdate is not null"),
+        new Index(
+            appendRandom("in_orgunitgroupsetstructure_null"),
+            IndexType.BTREE,
+            Unique.NON_UNIQUE,
+            List.of("organisationunitid", "startdate"),
+            "startdate is null"));
+  }
+
+  @Override
+  public ResourceTableType getTableType() {
+    return ResourceTableType.ORG_UNIT_GROUP_SET_STRUCTURE;
   }
 
   @Override
   public Optional<String> getPopulateTempTableStatement() {
     String sql =
         "insert into "
-            + getTempTableName()
+            + toStaging(TABLE_NAME)
             + " "
             + "select ou.organisationunitid as organisationunitid, ou.name as organisationunitname, null as startdate, ";
```

Comment:
Is there a reason that this SQL is not part of the responsibilities of the builder?

Reply:
Good question. I wanted to make SqlBuilder general and multi-purpose. This logic is very specific and different for each resource table type.

I did think of having ResourceTable  extend Table, but these methods require variable input arguments so it got tricky. We can revisit this later for sure.